torch/nn/modules/module.py

```diff
@@ -254,28 +268,35 @@ def register_forward_hook(self, hook):
         return handle
 
     def __call__(self, *input, **kwargs):
-        for hook in self._forward_pre_hooks.values():
-            hook(self, input)
-        result = self.forward(*input, **kwargs)
-        for hook in self._forward_hooks.values():
-            hook_result = hook(self, input, result)
-            if hook_result is not None:
-                raise RuntimeError(
-                    "forward hooks should never return any values, but '{}'"
-                    "didn't return None".format(hook))
-        if len(self._backward_hooks) > 0:
-            var = result
-            while not isinstance(var, Variable):
-                if isinstance(var, dict):
-                    var = next((v for v in var.values() if isinstance(v, Variable)))
-                else:
-                    var = var[0]
-            grad_fn = var.grad_fn
-            if grad_fn is not None:
-                for hook in self._backward_hooks.values():
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        tracing_state = torch.jit.get_tracing_state(_first_var(input, kwargs))
```

Comment:
nit: I think it would be nicer to have a context manager for handling the scopes, which would take care of the `try`/`finally` for you.

Reply:
I like this, we should make it (optionally?) available to users too

Reply:
Yes, I think it would be good to make this available at some point, once we have converged the interface nicely.

Reply:
Done